### DISCUSSION TITLE
Relation field not saving in pivot table

### DISCUSSION THREAD

**Discussion by willardinteractive** [orchidsoftware/platform#1724] [Q&A] [answered]
Is the relation field supposed to save when you have an relationship table?

I have a relationship field setup on my employee edit screen:

`Relation::make('locations')
     ->fromModel(Location::class, 'name')
     ->multiple()
     ->title('Assign to location'),`

The dropdown correctly lists all the location entries but doesn't seem to save. I have a database pivot table called "location_employee" with columns for location_id and employee_id.

In my Employee model I have:

`public function locations()
    {
        return $this->belongsToMany(\App\Models\Location::class);
    }`

Is this supposed to automatically save in the pivot table (and I have something incorrect here) or do I have to create my own code for saving data in a pivot table?

Thanks!

**Comment by willardinteractive** [comment, marked as answer]
In case anyone has this challenge in the future, Here is how I got my data saving and pulling from the pivot table.

In my Employee edit screen file I needed to add the following to the createOrUpdate method after the initial save:

`$employee->locations($employee->id)->sync(optional($request->get('employee'))['locations']);`

In the same file where I define my form fields I used:

```
Relation::make('employee.locations.')
                        ->fromModel(Branch::class, 'name')
                        ->multiple()
                        ->title('Assign to Location(s)')
                        ->help('Assign to one or more location.'),
```

In the same file in my query method I added the following before the return. This loads my pivot table data to populate the form field with the store value(s).

`$employee->load(['locations']);`

In my Model file I have the relationship defined like this:

`public function locations()
    {
        return $this->belongsToMany(\App\Models\Location::class, 'location_employee', 'employee_id','location_id');
    }`

All of this allowed me to save the data from the form field and show the saved value. Hope this helps someone else in the future.

**Comment by Longhot** [comment]
Thank you, due to you my problem is solved!!!

I just had a little trouble understanding that 
relation::make('employee.locations.')
must have the same name as the relation.
Then it became easy, thanks

**Comment by LemonPie22** [comment]
I've been struggling with the same issue for a while now.

I'm trying to relate Products to Products in a many-to-many relationship via a pivot table called `product_component`.

I've defined this relationship in my Product model as:
```
public function parent_products(): BelongsToMany
{
    return $this->belongsToMany(Product::class,'product_component','component_id','product_id');
}

public function child_products(): BelongsToMany
{
    return $this->belongsToMany(Product::class, 'product_component', 'product_id', 'component_id');
}
```


I'm using Orchid CRUD and my ProductResource file has the following relation fields:
```
FieldsRelation::make('child_products.')
    ->fromModel(Product::class, 'name')
    ->multiple()
    ->title('Children'),
FieldsRelation::make('product.child_products.')
    ->fromModel(Product::class, 'name')
    ->multiple()
    ->title('Child products'),
```
They're both trying to do the same thing (display and update related products) but both fail in different ways:
1. The first field displays related products but ignores updates.
2. The second field doesn't display related products but any updates in this field are saved.

I need a field that does both!

My ProductResource file has the following functions:
```
    /**
     * Action to query one model
     *
     * @param Model $model
     *
     * @return Model $model
     */
    public function modelQuery(ResourceRequest $request, Model $model): Builder
    {
        $model->load(['child_products']);
        // $model->load(['users']);
        return $model->query();
    }

    /**
     * Action to create and update the model
     *
     * @param ResourceRequest $request
     * @param Model           $model
     */
    public function save(ResourceRequest $request, Model $model): void
    {
        $model->fill($request->all())->save();
        $model->child_products()->sync(optional($request->get('product'))['child_products']);
        // $model->users()->sync(optional($request->get('product'))['users']);
    }
```

Any clues?

**Reply by Estigy** [reply]
I think the problem is that both of your fields are called "child_products". One should be named "parent_products", right?

**Reply by LemonPie22** [reply]
No that's not it. Eventually I want separate fields for parent and child products but, in this case, I am experimenting with 2 different ways to show child products.

**Reply by Estigy** [reply]
I tried the match the same setup you have:
I have `piece.categories.` and I added a second field named only `categories.` which loads its entities from the same model.
The line from my answer below still does the correct thing (syncing the `categories` relation on the `$piece` model), and the second field correctly yields an array with the things I selected for this input field.

I see no reason why this should not work for you, it totally works correctly for me. Do you have the latest version of Orchid?

**Reply by LemonPie22** [reply]
@Estigy Yes I'm on version 12.6.2.

The only significant difference I can spot between your code and mine is that we use different variables in the save method. I use `$model` and you use `$piece`.

I posted the full code of my save method above.

`$model` is declared at the start of my ProductResource class:
```
class ProductResource extends Resource
{
    /**
     * The model the resource corresponds to.
     *
     * @var string
     */
    public static $model = \App\Models\Product::class;
```

Should I be using another variable rather than `$model`? Maybe something like `$product`?

**Reply by Estigy** [reply]
Well, it indeed seems we are talking about different things here.
I'm using the Orchid platform, no crud resources at work here.

**Reply by LemonPie22** [reply]
Yup, I'm using Orchid CRUD. It seemed like a good idea at the time 🤣

**Reply by LemonPie22** [reply]
@Estigy @tabuna, is the conclusion to this that relation doesn't work with Orchid CRUD? I'm trying to decide if I should keep trying with CRUD or switch to screens.

**Reply by LemonPie22** [reply]
Never mind, I've **finally** solved it!

I used this way of defining the relation:
```
FieldsRelation::make('child_products.')
    ->fromModel(Product::class, 'name')
    ->multiple()
    ->title('Child Products'),
```

`Select` also works (not sure what the difference is with `relation`), e.g:
```
Select::make('users.')
    ->fromModel(User::class, 'name')
    ->multiple()
    ->title('SMEs')
    ->help('Subject Matter Experts'),
```
Then all that's needed is a `save` function in the resource class like this:
```
/**
 * Action to create and update the model
 *
 * @param ResourceRequest $request
 * @param Model           $model
 */
public function save(ResourceRequest $request, Model $model): void
{
    $model->fill($request->all())->save();

    // child products
    if (isset($request->request->all()['child_products'])) {
        $model->child_products()->sync($request->request->all()['child_products']);
    }
    else
    {
        $model->child_products()->detach();
    }

    // subject matter experts
    if (isset($request->request->all()['users'])) {
        $model->users()->sync($request->request->all()['users']);
    }
    else
    {
        $model->users()->detach();
    }

}
```
```

**Comment by Estigy** [comment]
@tabuna 
Can you confirm that the selected answer here actually is the correct answer?
I would then send you a pull request with the updated documentation :-)

My solution was to add one line to the save method:
`$piece->categories()->sync(optional($request->get('piece'))['categories']);`
And I did not need to add anything to get the form field populated correctly.

I think a line like this should be added to the documentation. (Or are we wrong and the relation field should indeed be updated automatically, and there is a bug somewhere prohibiting this?)

**Reply by LemonPie22** [reply]
I added a very similar line to my save method:
`$model->child_products()->sync(optional($request->get('product'))['child_products']);`
But it doesn't work for me @Estigy
Can you see what I'm doing wrong?

**Reply by Estigy** [reply]
Did you add the dot at the end of the fieldname? It should read "product.child_products." with a dot at the end.

**Reply by Estigy** [reply]
Please see my reply further above.

**Reply by LemonPie22** [reply]
Thanks for jumping in on this @Estigy. I appreciate your help.
Yes I added a dot. I posted my exact code above.
